pkg/utils/apply/patch.go

```diff
@@ -0,0 +1,160 @@
+package apply
+
+import (
+	"encoding/json"
+
+	"github.com/oam-dev/kubevela/pkg/oam"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/jsonmergepatch"
+	"k8s.io/apimachinery/pkg/util/mergepatch"
+	"k8s.io/apimachinery/pkg/util/strategicpatch"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var metadataAccessor = meta.NewAccessor()
+
+// threeWayMergePatch creates a patch by computing a three way diff.
+func threeWayMergePatch(currentObj, modifiedObj runtime.Object) (client.Patch, error) {
+	current, err := json.Marshal(currentObj)
+	if err != nil {
+		return nil, err
+	}
+	original, err := getOriginalConfiguration(currentObj)
```

Comment:
what will happen if the original is nil?

Reply:
if original is nil, three-way-merge (both StrategicMerge and JSONMerge) will use "{}" as the original state. It would not result in a panic or failure.